src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFileImpl.java

```diff
@@ -0,0 +1,133 @@
+package edu.kpi.testcourse.storage;
+
+import com.google.gson.reflect.TypeToken;
+import edu.kpi.testcourse.entities.UrlAlias;
+import edu.kpi.testcourse.logic.UrlShortenerConfig;
+import edu.kpi.testcourse.serialization.JsonTool;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * A file-backed implementation of {@link UrlRepository} suitable for use in production.
+ */
+public class UrlRepositoryFileImpl implements UrlRepository {
+
+  // Urls, keyed by alias.
+  private final Map<String, UrlAlias> urlMapByAlias;
+  // All User urls, keyed by email
+  private final Map<String, List<UrlAlias>> urlsMapByEmail;
+
+  private final JsonTool jsonTool;
+  private final Path jsonFilePath;
+
+  /**
+   * Creates an instance.
+   */
+  @Inject
+  public UrlRepositoryFileImpl(JsonTool jsonTool, UrlShortenerConfig appConfig) {
+    this.jsonTool = jsonTool;
+    this.jsonFilePath = makeJsonFilePath(appConfig.storageRoot());
+    this.urlMapByAlias = readUrlsFromJsonDatabaseFile(jsonTool, this.jsonFilePath);
+    this.urlsMapByEmail = makeUrlsMapByEmail(urlMapByAlias);
+  }
+
+  @Override
+  public synchronized void createUrlAlias(UrlAlias urlAlias) throws AliasAlreadyExist {
```

Comment:
We must avoid hashmap changes from another thread while the method execution to have the data in a consistent state. The queue will not help us because another thread can call another method from UrlRepository and change the data state. So now the UrlRepository is a Monitor that provides access to the hashmap.

Reply:
Good old Korochkin's lectons, now I see